### PR TITLE
Rancher 999

### PR DESCRIPTION
**Pull request by guram-jalaghonia**
folioCIQualityGate Job
Compare Commit Hash 
Previous Commit ~Hash Saved TO AWS SSM   
https://issues.folio.org/browse/RANCHER-999
https://jenkins-aws.indexdata.com/job/folioRancher/job/folioQualityGates/job/folioCiQualityGates-RANCHER-1000/73/

### CODE REVIEW

File: pipelines/folioRancher/folioQualityGates/Jenkinsfile [hunk 1]
```diff
@@ -27,20 +29,36 @@ pipeline {
...
-                    if (params.SKIP_CHANGES == true) {
-                        echo "Skipping changes For the SNAPSHOT BRANCH and moving to Parrarel Run ."
-                    } else {
-                        checkout scm
-                        echo "Checking Changes for the SNAPSHOT BRANCH..and moving to Parrarel Run ."
-                    }
+stage('Check Changes') {
+    steps {
+        script {
+           if (params.SKIP_CHANGES) {
```
Comment (OHaimanov) [inline pipelines/folioRancher/folioQualityGates/Jenkinsfile:35]: Maybe better to name this param like 'SKIP_CHECK'?

File: pipelines/folioRancher/folioQualityGates/Jenkinsfile [hunk 2]
```diff
@@ -27,20 +29,36 @@ pipeline {
...
+                                  secretKeyVariable: 'AWS_SECRET_ACCESS_KEY']]) {
+
+                def HashDiffDetected = folioHashCommitCheck.commitHashChangeDetected(params.BRANCH)  // true or false
+                echo "HashCommitChanges: ${HashDiffDetected}"
+                env.HashDiffDetected = HashDiffDetected.toString()
+                if (HashDiffDetected) {
+                   // Set the build status to UNSTABLE (Yellow) To Indicate that Changes Are Detected, For the Visualisation Purposes
+                   echo "Changes detected, setting build status to UNSTABLE"
+                   currentBuild.result = 'UNSTABLE'
+                   }
```
Comment (OHaimanov) [inline pipelines/folioRancher/folioQualityGates/Jenkinsfile:50]: I think that this indication is not necessary

File: pipelines/folioRancher/folioQualityGates/Jenkinsfile [hunk 3]
```diff
@@ -27,20 +29,36 @@ pipeline {
...
+                echo "Skipping changes for the SNAPSHOT BRANCH and moving to Parallel Run."
+            } else {
+                withCredentials([[$class           : 'AmazonWebServicesCredentialsBinding',
+                                  credentialsId    : Constants.AWS_CREDENTIALS_ID,
+                                  accessKeyVariable: 'AWS_ACCESS_KEY_ID',
+                                  secretKeyVariable: 'AWS_SECRET_ACCESS_KEY']]) {
+
+                def HashDiffDetected = folioHashCommitCheck.commitHashChangeDetected(params.BRANCH)  // true or false
+                echo "HashCommitChanges: ${HashDiffDetected}"
+                env.HashDiffDetected = HashDiffDetected.toString()
```
Comment (OHaimanov) [inline pipelines/folioRancher/folioQualityGates/Jenkinsfile:45]: I think that better to move withCredentials block to awscli.groovy and create wrapper (closure) like folioHelm.withK8sClient

Comment (OHaimanov) [inline pipelines/folioRancher/folioQualityGates/Jenkinsfile:45]: Also env. variables better to name in CAPITAL_CASE

File: pipelines/folioRancher/folioQualityGates/Jenkinsfile [hunk 4]
```diff
@@ -27,20 +29,36 @@ pipeline {
...
+                if (HashDiffDetected) {
+                   // Set the build status to UNSTABLE (Yellow) To Indicate that Changes Are Detected, For the Visualisation Purposes
+                   echo "Changes detected, setting build status to UNSTABLE"
+                   currentBuild.result = 'UNSTABLE'
```
Comment (OHaimanov) [inline pipelines/folioRancher/folioQualityGates/Jenkinsfile:49]: So i think that clear code here should looks like
env.DIFF_DETECTED=folioHashCommitCheck.commitHashChangeDetected(params.BRANCH).toString()

File: vars/folioHashCommitCheck.groovy [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+  def previousSavedHash = getPreviousSavedHashFromSSM(awsRegion,awsParameterName)
+  echo "latestCommitHash ${latestCommitHash}"
+  echo "previousSavedHash ${previousSavedHash}"
+
+  if (latestCommitHash == previousSavedHash) {
+    echo "No changes detected. HashDiffDetected :false."
+    return false
+  } else {
+    echo "Changes detected. HashDiffDetected Returning true and updating ssm with new hash: ${latestCommitHash}."
+    withUpdateSsmParameterNewValue(awsRegion,awsParameterName,latestCommitHash)
```
Comment (OHaimanov) [inline vars/folioHashCommitCheck.groovy:22]: wrap this with closure awscli.withClient{} or awscli.withCreds{}

File: vars/folioHashCommitCheck.groovy [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    error "An error occurred while fetching GitHub data: ${e.message}"
+    return null
+  }
+}
+
+// Function to get the previous platforme-complete $branch commit saved hash from AWS SSM
+String getPreviousSavedHashFromSSM(awsRegion, awsParameterName) {
+  try {
+    def parameterValue = sh(
+      script: "aws ssm get-parameter --name ${awsParameterName} --region ${awsRegion} --query 'Parameter.Value' --output text",
```
Comment (OHaimanov) [inline vars/folioHashCommitCheck.groovy:55]: Maybe it is better to move this code to awscli like getSsmParameter?

File: vars/folioHashCommitCheck.groovy [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+  } catch (Exception e) {
+    error "Error fetching parameter value from AWS SSM: ${e.message}"
+    return null
+  }
+}
+
+
+// Function to update the SSM parameter with a new hash value
+void withUpdateSsmParameterNewValue(String awsRegion, String awsParameterName, String latestCommitHash) {
+  sh(script: "aws ssm put-parameter --name ${awsParameterName} --region ${awsRegion} --value ${latestCommitHash} --type String --overwrite")
```
Comment (OHaimanov) [inline vars/folioHashCommitCheck.groovy:69]: Maybe it is better to move this code to awscli like updateSsmParameter?

**Review by OHaimanov — CHANGES_REQUESTED** [review]

File: pipelines/folioRancher/folioQualityGates/Jenkinsfile (hunk 1 above)
Comment (guram-jalaghonia) [reply pipelines/folioRancher/folioQualityGates/Jenkinsfile:35]: Adjusted, in new commit will be included

File: pipelines/folioRancher/folioQualityGates/Jenkinsfile (hunk 4 above)
Comment (guram-jalaghonia) [reply pipelines/folioRancher/folioQualityGates/Jenkinsfile:49]: thanks , good one ,i will use like this

File: vars/awscli.groovy [hunk 8]
```diff
@@ -65,3 +67,30 @@ void waitRdsClusterAvailable(String cluster_name, String region) {
...
+ */
+void updateSsmParameter(String awsRegion, String awsParameterName, String latestCommitHash) {
+  sh(script: "aws ssm put-parameter --name ${awsParameterName} --region ${awsRegion} --value ${latestCommitHash} --type String --overwrite")
+}
+
+/**
+ * @param awsRegion The aws region where located ssm parameter
+ * @param awsParameterName parameter where we are saving the commit hash value
+ */
+String GetSsmParameterValue(String awsRegion, String awsParameterName) {
```
Comment (OHaimanov) [inline vars/awscli.groovy:84]: rename Get to get

**Review by OHaimanov — APPROVED** [review]

**Pull request merged**